doc/developer/design/20220314_linearized_reads.md

```diff
@@ -0,0 +1,160 @@
+
+True _linearizability_ requires that reads and writes happen in some _total order_ that is
+consistent with the _real time ordering_ of those operations. That is, each operation *appears*
+to happen atomically at some singular point in time bounded by the beginning and end of that
+operation.
+
+A `LINEARIZABLE` read against some view `V` is possible if-and-only-if the following are true:
+
+* All _source_ pTVC's that are at the base of `V` offer a _linearizable_ API for obtaining the
+  latest offset of data for that pTVC.
+  * "offset" here refers to some monotonically increasing identifier that durably and consistently
+  labels data.
+  * A given API is _linearizable_ if-and-only-if later (as in, later in _real-time_) calls tho
+  that API can never produce lower "offets".
+* For all _source_ pTVC's involved, materialized durably associate _real-time_ timestamps
+with specific offsets (these are typically referred to as _timestamp bindings_ and are currently
+implemented for Kafka sources).
+* ...and the responses from those API's can be given some _real-time total order_.
+  * The simplest way to obtain this constraint is to involve only one source.
+
+With the above constraints, the mechanism that materialized uses to service the read
+against some view `V` is as follows:
+
+1. Fetch the latest offsets for all _source_ pTVC's involved.
+2. Wait for timestamps to be durably recorded associated with all offsets, as `source_timestamps`
+3. Choose a timestamp that is _at least_
+`max(max(source_uppers), max(source_timestamps))`. Note that this may require one of both of
+holding back the `since` of `V` or waiting for the `upper` to progress past the chosen timestamp.
+4. Service the read for the view at that timestamp. This will durably advance the `upper`s of all involved
+_source_ pTVC's to at least that timestamp.
+
+This provides _linearizability_ because:
+* materialized guarantees (possible with a timestamp oracle service) that all calls to
+`current_real_time()` monotonically increase.
+* Durably recoding offsets associated with timestamps means that reads (even non-linearizable ones)
+at later times can violate the _real-time_ ordering, as later reads will always have
+a _real-time_ timestamp that is at least the chosen timestamp, and materialized, for a read against
+a pTVC at some timestamp `t` always includes all data with offsets whose associated timestamps are <= `t`.
+
+The reason the last constraint above is important is because this process can create
+an inviolably _real-time_ ordering between writes the upstream source and reads from
+materialized views, without that constraint materialized has no information if offsets
+from disparate _sources_ have a causal relationship, and could thereforce produce
+results
+
+### Caveat: views involving tables
+All reads against views that involve tables will force the query to happen at a global
+`linearized` timestamp that is shared across tables. This may be greater than the
```

Comment:
What if it is less than?

Reply:
we choose the higher one? 
will the table query me mad about that?

Reply:
Nah the query itself is fine. Just thinking through if it's still linearizable if it does that. I think it's ok because this just changes the return of determine_timestamp, which is one input to the final timestamp.